### PR TITLE
[CALCITE-6527] Add DATE_ADD function (enabled in Spark library)

### PR DESCRIPTION
**Pull request by caicancai**
https://issues.apache.org/jira/browse/CALCITE-6527

### CODE REVIEW

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 1]
```diff
@@ -13824,6 +13824,54 @@ void testTimestampDiff(boolean coercionEnabled) {
...
+        "2016-02-20",
+        "DATE NOT NULL");
+    f.checkScalar("date_add(timestamp '2016-02-22 13:00:01', -2)",
+        "2016-02-20",
+        "DATE NOT NULL");
+    f.checkScalar("date_add(timestamp '2016-02-22 13:00:01', -2.0)",
+        "2016-02-20",
+        "DATE NOT NULL");
+    f.checkNull("date_add(CAST(NULL AS DATE), 5)");
+    f.checkNull("date_add(date '2016-02-22', CAST(NULL AS INTEGER))");
```
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13871]: how about a few tests around year 0 and the Gregorian calendar cutoff date?
For example, date_add('0001-01-01', -2). Notice that there is no year 0, so this probably should throw.

Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13871]: Good suggestion. I tested it in spark and it seems that spark is not aware of this error. Should we fix this error in calcite or keep this behavior?
scala> val df = spark.sql("SELECT date_add('0001-01-01', -2)")
df: org.apache.spark.sql.DataFrame = [date_add(0001-01-01, -2): date]

```
scala> df.show()
+------------------------+
|date_add(0001-01-01, -2)|
+------------------------+
|              0000-12-30|
+------------------------+


scala> val df = spark.sql("SELECT date_add('0000-01-01', -2)")
df: org.apache.spark.sql.DataFrame = [date_add(0000-01-01, -2): date]

scala> df.show()
+------------------------+
|date_add(0000-01-01, -2)|
+------------------------+
|             -0001-12-30|
+------------------------+

```

Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13871]: You should check the issues open for spark, this looks like a bug to me. There is no year 0.
If there is no bug maybe you can open one.

Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13871]: OK, I will try again in the latest spark code. If there is no corresponding jira record, I will file a jira case to describe this bug.

I feel that 000-00-01 is reasonable, but negative numbers are unreasonable. Calcite currently follows this principle. The following are my test results in Calcite.
```
    f.checkScalar("date_add('0001-01-01', -2)",
        "0000-12-30",
        "DATE NOT NULL");
    f.checkFails("date_add('0000-01-01', -2)",
        "Value of YEAR field is out of range in '0000-01-01'",
        true);

```

Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13871]: on the contrary, year 0 is not legal, it never existed.
and people lived before Christ, so years BC are completely legal in many databases. whether they show up as negative numbers, that's a different question. Calcite doesn't support them, though

File: site/_docs/reference.md [hunk 2]
```diff
@@ -2772,6 +2772,7 @@ In the following:
 | b s | DATE_FROM_UNIX_DATE(integer)                 | Returns the DATE that is *integer* days after 1970-01-01
 | p r | DATE_PART(timeUnit, datetime)                | Equivalent to `EXTRACT(timeUnit FROM  datetime)`
 | b | DATE_ADD(date, interval)                       | Returns the DATE value that occurs *interval* after *date*
+| s | DATE_ADD(start_date, num_days)                 | Returns the DATE that is *num_days* after *start_date*
```
[the remark this answers is not in the document]
Comment (caicancai) [reply site/_docs/reference.md:2775]: Thanks for the reminder, I'll fix it later.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 1 above)
Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13871]: I have already raised a jira case in spark
https://issues.apache.org/jira/browse/SPARK-49247

File: site/_docs/reference.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (caicancai) [reply site/_docs/reference.md]: done

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 1 above)
Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13871]: I think that in my previous review I asked for a few more tests around boundary conditions, I don't see any here.

Comment (julianhyde) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13871]: _Au contraire_. `0000-00-00` is a valid 8601 date.

See https://en.wikipedia.org/wiki/ISO_8601#Calendar_dates

Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13871]: Will the Calcite DATE datatype accept such a value?

Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13871]: I think I will do some tests and describe them later. It seems that calcite has done some processing but it is not perfect. I will fix it in this PR

Comment (julianhyde) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13871]: @mihaibudiu, I don't recall. I just happened to reading about ISO-8601 so I thought I'd correct misinformation.

Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13880]: Sorry for the late reply.
After testing, the date_add function will format the date. For example, invalid dates such as `0000-01-01` will fail the date format.

**Comment by caicancai** [comment]
@mihaibudiu I added some tests that I thought were borderline and found some issues, I'd like to hear your thoughts

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 3]
```diff
@@ -13824,6 +13824,63 @@ void testTimestampDiff(boolean coercionEnabled) {
...
+    f.checkScalar("date_add(timestamp '2016-02-22 13:00:01', -2)",
+        "2016-02-20",
+        "DATE NOT NULL");
+    f.checkScalar("date_add(timestamp '2016-02-22 13:00:01', -2.0)",
+        "2016-02-20",
+        "DATE NOT NULL");
+    f.checkScalar("date_add(date '0001-01-01', '-2.0')",
+        "0000-12-30",
+        "DATE NOT NULL");
+    f.checkScalar("date_add(date '0001-01-01', '-367')",
```
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13873]: what is going on here? Why is `000/` a valid year?

Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13873]: @mihaibudiu I raised a jira to discuss this issue. Should I delete this test?
https://issues.apache.org/jira/browse/CALCITE-6536

Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13873]: I would still like to know why this test passes.

Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13873]: If a test passes that shouldn't it means that there's somewhere another bug that is masking a problem.

Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13873]: > If a test passes that shouldn't it means that there's somewhere another bug that is masking a problem.

agree. 
I will fix it, the date function does validation when it handles 000/or 0000 as a parameter, but when 000/or 0000 is returned as a result, no such validation is done, I think this is the bug

Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13873]: @mihaibudiu I'm not sure how to solve this bug, it looks a bit complicated to me 😭

Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13873]: I don't think Calcite can offer exact compatibility to Spark in this case.
The best solution for Calcite is to give a runtime error.
But I would still like to know why the date with a slash didn't report a different error.

Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13873]: > I don't think Calcite can offer exact compatibility to Spark in this case.
> The best solution for Calcite is to give a runtime error.
> But I would still like to know why the date with a slash didn't report a different error.

@mihaibudiu Calcite will perform parser validation on the input parameters.
```
    final Calendar cal =
        DateTimeUtils.parseDateFormat(s, Format.get().date,
            DateTimeUtils.UTC_ZONE);
```
However, there is no restriction on the output result set. I've been researching for a long time and I'm not sure where to limit the result set

Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13873]: I don't understand what you are saying.
If you add two very large numbers in Calcite you will get a runtime error.
How is this different?

Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13873]: I mean, I don't know why I get a runtime error.
Sorry, I didn't express myself clearly.

**Comment by caicancai** [comment]
@mihaibudiu Almost all date functions have this problem, I consider it another bug, I will fix it. Before fixing it, this PR is marked as draft.
Thank you for your review
